src/session/SessionManager.cpp

```diff
@@ -0,0 +1,303 @@
+        auto spaceName = session.get_space_name();
+        SpaceInfo spaceInfo;
+        if (!spaceName.empty()) {
+            VLOG(2) << "Get session with spaceName: " << spaceName;
+            // get spaceInfo
+            auto spaceId = metaClient_->getSpaceIdByNameFromCache(spaceName);
+            if (!spaceId.ok()) {
+                LOG(ERROR) << "Get session with Unknown space: " << spaceName;
+            } else {
+                auto spaceDesc = metaClient_->getSpaceDesc(spaceId.value());
+                if (!spaceDesc.ok()) {
+                    LOG(ERROR) << "Get session with Unknown space: " << spaceName;
+                }
+                spaceInfo.id = spaceId.value();
+                spaceInfo.name = spaceName;
+                spaceInfo.spaceDesc = std::move(spaceDesc).value();
+            }
+        }
+
+        folly::RWSpinLock::WriteHolder wHolder(rwlock_);
+        auto findPtr = activeSessions_.find(id);
+        if (findPtr == activeSessions_.end()) {
+            VLOG(1) << "Add session id: " << id << " from metad";
+            auto sessionPtr = ClientSession::create(std::move(session), metaClient_);
+            sessionPtr->charge();
+            activeSessions_[id] = sessionPtr;
+
+            // update the space info to sessionPtr
+            if (!spaceName.empty()) {
+                sessionPtr->setSpace(std::move(spaceInfo));
+            }
+
+            return sessionPtr;
+        }
+        return findPtr->second;
+    };
+
+
+    auto execInstance = new ExecuteInstance(std::move(rctx));
+    auto doFinish = [this, execInstance, execFunc](auto&& resp) {
+        VLOG(2) << "Find session doFinish";
+        if (!resp.ok()) {
+            execInstance->rctx()->resp().errorCode = ErrorCode::E_SESSION_INVALID;
+            execInstance->rctx()->resp().errorMsg.reset(new std::string("Session not found."));
+            execInstance->rctx()->finish();
+            return execInstance->finish();
+        }
+        auto session = std::move(resp).value();
+        session->updateGraphAddr(myAddr_);
+        auto roles = metaClient_->getRolesByUserFromCache(session->user());
+        for (const auto &role : roles) {
+            session->setRole(role.get_space_id(), role.get_role_type());
+        }
+        execInstance->rctx()->setSession(std::move(session));
+        execFunc(execInstance->moveRctx());
+        return execInstance->finish();
+    };
+
+    metaClient_->getSession(id).via(runner).thenValue(addSession).thenValue(doFinish);
+}
+
+void SessionManager::createSession(const std::string &userName,
+                                   const std::string& clientIp,
+                                   folly::Executor* runner,
+                                   std::unique_ptr<RequestContext<AuthResponse>> rctx) {
+    {
+        folly::RWSpinLock::ReadHolder rHolder(rwlock_);
+        if (activeSessions_.size() >= FLAGS_max_allowed_connections) {
+            return rctx->finish();
```

Comment:
I think the `SessionManager` has no responsibility to finish the `RequestContext`. And we should return error here.

Reply:
good catch